src/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -0,0 +1,222 @@
+-- | A general-purpose fence pointer index.
+module Database.LSMTree.Internal.IndexOrdinary
+(
+    IndexOrdinary (IndexOrdinary),
+
+    -- * Search
+    search,
+
+    -- * Deserialisation
+    fromSBS
+)
+where
+
+import Prelude hiding (length, drop, takeWhile)
+
+import Control.Monad
+       (
+           when
+       )
+import Data.Word
+       (
+           Word8,
+           Word16,
+           Word32,
+           Word64,
+           byteSwap32
+       )
+import Data.Primitive.ByteArray
+       (
+           ByteArray,
+           indexByteArray
+       )
+import Data.ByteString.Short
+       (
+           ShortByteString (ShortByteString)
+       )
+import Data.ByteString.Short qualified as ShortByteString
+       (
+           length
+       )
+import Data.Vector
+       (
+           Vector,
+           fromList,
+           length,
+           (!),
+           drop,
+           takeWhile,
+           findIndexR
+       )
+import Data.Vector.Primitive qualified as Primitive
+       (
+           Vector (Vector),
+           force,
+           null,
+           length,
+           take,
+           drop,
+           splitAt,
+       )
+import Database.LSMTree.Internal.Serialise
+       (
+           SerialisedKey (SerialisedKey')
+       )
+import Database.LSMTree.Internal.Entry
+       (
+           NumEntries (NumEntries)
+       )
+import Database.LSMTree.Internal.IndexCompact
+       (
+           PageNo (PageNo),
+           PageSpan (PageSpan)
+       )
+
+{-|
+    The version of the index and its serialisation format that is supported by
+    this module.
+-}
+supportedVersion :: Word32
+supportedVersion = 1
+
+{-|
+    A general-purpose fence pointer index.
+
+    An index is represented by a vector that maps the number of each page to the
+    key stored last in this page or, if the page is an overflow page, to the key
+    of the corresponding key–value pair. From this, the following constraints on
+    the vector follow:
+
+      * The vector must be non-empty.
+
+      * The elements of the vector must be non-decreasing.
+-}
+newtype IndexOrdinary = IndexOrdinary (Vector SerialisedKey)
+
+-- * Search
+
+{-|
+    Returns the span of pages that contain the key–value pair of a certain key
+    if there is such a pair, or an arbitrary, valid page span if there is not.
+-}
+search :: SerialisedKey -> IndexOrdinary -> PageSpan
+search key (IndexOrdinary lastKeys) = PageSpan (PageNo start) (PageNo end) where
+
+    start :: Int
+    start | protoStart < pageCount = protoStart
+          | otherwise              = maybe 0 succ $
+                                     findIndexR (/= lastKeys ! pred pageCount) $
+                                     lastKeys
+
+    end :: Int
+    end | protoStart < pageCount
+            = start + length (takeWhile (== key) (drop (succ start) lastKeys))
+        | otherwise
+            = pred pageCount
```

Comment:
It looks to me like this could return `-1`

Reply:
Yes, it could, but only if the run contains no pages, which should be disallowed anyhow, because no sensible page span can be returned in this case (as also argued in [my comment on the documentation of `search`](https://github.com/IntersectMBO/lsm-tree/pull/244#discussion_r1636656680)).

That said, it might not be a good idea to silently return a page span from 0 to −1 in case of no pages. Maybe we should add an assertion that the length of the vector representing the index is not zero somewhere, perhaps at the outside of the right-hand side of the `search` definition. What is your opinion on this?

Reply:
An `assert` could be useful, yes

Reply:
Added in da5e3a80a45d590999e9e11c6f8c5de215b00319.